In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_csv('survey_results_public.csv')

In [3]:
df.head()

ResponseId                                         MainBranch  \
0           1                                      None of these   
1           2                     I am a developer by profession   
2           3  I am not primarily a developer, but I write co...   
3           4                     I am a developer by profession   
4           5                     I am a developer by profession   

            Employment                            RemoteWork  \
0                  NaN                                   NaN   
1  Employed, full-time                          Fully remote   
2  Employed, full-time  Hybrid (some remote, some in-person)   
3  Employed, full-time                          Fully remote   
4  Employed, full-time  Hybrid (some remote, some in-person)   

                           CodingActivities  \
0                                       NaN   
1  Hobby;Contribute to open-source projects   
2                                     Hobby   
3              I don’t code outside of work   
4                                     Hobby   

                                           EdLevel  \
0                                              NaN   
1                                              NaN   
2  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
3     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                           LearnCode  \
0                                                NaN   
1                                                NaN   
2  Books / Physical media;Friend or family member...   
3  Books / Physical media;School (i.e., Universit...   
4  Other online resources (e.g., videos, blogs, f...   

                                     LearnCodeOnline LearnCodeCoursesCert  \
0                                                NaN                  NaN   
1                                                NaN                  NaN   
2  Technical documentation;Blogs;Programming Game...                  NaN   
3                                                NaN                  NaN   
4  Technical documentation;Blogs;Stack Overflow;O...                  NaN   

  YearsCode YearsCodePro                                            DevType  \
0       NaN          NaN                                                NaN   
1       NaN          NaN                                                NaN   
2        14            5  Data scientist or machine learning specialist;...   
3        20           17                              Developer, full-stack   
4         8            3  Developer, front-end;Developer, full-stack;Dev...   

                OrgSize      PurchaseInfluence  \
0                   NaN                    NaN   
1                   NaN                    NaN   
2    20 to 99 employees  I have some influence   
3  100 to 499 employees  I have some influence   
4    20 to 99 employees  I have some influence   

                                          BuyNewTool  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                            Other (please specify):   
4  Start a free trial;Visit developer communities...   

                                             Country  \
0                                                NaN   
1                                             Canada   
2  United Kingdom of Great Britain and Northern I...   
3                                             Israel   
4                           United States of America   

                    Currency  CompTotal CompFreq  \
0                        NaN        NaN      NaN   
1       CAD\tCanadian dollar        NaN      NaN   
2        GBP\tPound sterling    32000.0   Yearly   
3    ILS\tIsraeli new shekel    60000.0  Monthly   
4  USD\tUnited States dollar        NaN      NaN   

                        LanguageHaveWo

In [4]:
df.shape

(73268, 79)

### Observation
- our data consists of 73268 rows and 79 columns.
- to display all the columns--

In [5]:
#Let's start work with columns
df.columns

Index(['ResponseId', 'MainBranch', 'Employment', 'RemoteWork',
       'CodingActivities', 'EdLevel', 'LearnCode', 'LearnCodeOnline',
       'LearnCodeCoursesCert', 'YearsCode', 'YearsCodePro', 'DevType',
       'OrgSize', 'PurchaseInfluence', 'BuyNewTool', 'Country', 'Currency',
       'CompTotal', 'CompFreq', 'LanguageHaveWorkedWith',
       'LanguageWantToWorkWith', 'DatabaseHaveWorkedWith',
       'DatabaseWantToWorkWith', 'PlatformHaveWorkedWith',
       'PlatformWantToWorkWith', 'WebframeHaveWorkedWith',
       'WebframeWantToWorkWith', 'MiscTechHaveWorkedWith',
       'MiscTechWantToWorkWith', 'ToolsTechHaveWorkedWith',
       'ToolsTechWantToWorkWith', 'NEWCollabToolsHaveWorkedWith',
       'NEWCollabToolsWantToWorkWith', 'OpSysProfessional use',
       'OpSysPersonal use', 'VersionControlSystem', 'VCInteraction',
       'VCHostingPersonal use', 'VCHostingProfessional use',
       'OfficeStackAsyncHaveWorkedWith', 'OfficeStackAsyncWantToWorkWith',
       'OfficeStackSyncHaveWork

1. Numbered item
1. Numbered item

1. Numbered item
     1. Substep

#### we can see that our column name consits of camel case letters in the first step we will try to convert all the column names in smaller letters

In [6]:
#now we will remove the _ in the column name
df.columns=df.columns.str.replace('_', '', regex=True)

In [7]:
df.columns=df.columns.str.lower()

In [8]:
df.head(2)

responseid                      mainbranch           employment  \
0           1                   None of these                  NaN   
1           2  I am a developer by profession  Employed, full-time   

     remotework                          codingactivities edlevel learncode  \
0           NaN                                       NaN     NaN       NaN   
1  Fully remote  Hobby;Contribute to open-source projects     NaN       NaN   

  learncodeonline learncodecoursescert yearscode yearscodepro devtype orgsize  \
0             NaN                  NaN       NaN          NaN     NaN     NaN   
1             NaN                  NaN       NaN          NaN     NaN     NaN   

  purchaseinfluence buynewtool country              currency  comptotal  \
0               NaN        NaN     NaN                   NaN        NaN   
1               NaN        NaN  Canada  CAD\tCanadian dollar        NaN   

  compfreq languagehaveworkedwith languagewanttoworkwith  \
0      NaN                    NaN                    NaN   
1      NaN  JavaScript;TypeScript        Rust;TypeScript   

  databasehaveworkedwith databasewanttoworkwith platformhaveworkedwith  \
0                    NaN                    NaN                    NaN   
1                    NaN                    NaN                    NaN   

  platformwanttoworkwith webframehaveworkedwith webframewanttoworkwith  \
0                    NaN                    NaN                    NaN   
1                    NaN                    NaN                    NaN   

  misctechhaveworkedwith misctechwanttoworkwith toolstechhaveworkedwith  \
0                    NaN                    NaN                     NaN   
1                    NaN                    NaN                     NaN   

  toolstechwanttoworkwith newcollabtoolshaveworkedwith  \
0                     NaN                          NaN   
1                     NaN                          NaN   

  newcollabtoolswanttoworkwith opsysprofessional use  \
0                          NaN                   NaN   
1                          NaN                 macOS   

                   opsyspersonal use versioncontrolsystem vcinteraction  \
0                                NaN                  NaN           NaN   
1  Windows Subsystem for Linux (WSL)                  Git           NaN   

   vchostingpersonal use  vchostingprofessional use  \
0                    NaN                        NaN   
1                    NaN                        NaN   

  officestackasynchaveworkedwith officestackasyncwanttoworkwith  \
0                            NaN                            NaN   
1                            NaN                            NaN   

  officestacksynchaveworkedwith officestacksyncwanttoworkwith  \
0                           NaN                           NaN   
1                           NaN                           NaN   

         blockchain                                         newsosites  \
0               NaN                                                NaN   
1  Very unfavorable  Collectives on Stack Overflow;Stack Overflow f...   

             sovisitfreq soaccount             sopartfreq    socomm  age  \
0                    NaN       NaN                    NaN       NaN  NaN   
1  Daily or almost daily       Yes  Daily or almost daily  Not sure  NaN   

  gender trans sexuality ethnicity accessibility mentalhealth tbranch icorpm  \
0    NaN   NaN       NaN       NaN           NaN          NaN     NaN    NaN   
1    NaN   NaN       NaN       NaN           NaN          NaN      No    NaN   

   workexp knowledge1 knowledge2 knowledge3 knowledge4 knowledge5 knowledge6  \
0      NaN        NaN        NaN        NaN        NaN        NaN        NaN   
1      NaN        NaN        NaN        NaN        NaN        NaN        NaN   

  knowledge7 frequency1 frequency2 frequency3 timesearching timeanswering  \
0        NaN        NaN        NaN        NaN           NaN           NaN   
1        

In [9]:
df.columns

Index(['responseid', 'mainbranch', 'employment', 'remotework',
       'codingactivities', 'edlevel', 'learncode', 'learncodeonline',
       'learncodecoursescert', 'yearscode', 'yearscodepro', 'devtype',
       'orgsize', 'purchaseinfluence', 'buynewtool', 'country', 'currency',
       'comptotal', 'compfreq', 'languagehaveworkedwith',
       'languagewanttoworkwith', 'databasehaveworkedwith',
       'databasewanttoworkwith', 'platformhaveworkedwith',
       'platformwanttoworkwith', 'webframehaveworkedwith',
       'webframewanttoworkwith', 'misctechhaveworkedwith',
       'misctechwanttoworkwith', 'toolstechhaveworkedwith',
       'toolstechwanttoworkwith', 'newcollabtoolshaveworkedwith',
       'newcollabtoolswanttoworkwith', 'opsysprofessional use',
       'opsyspersonal use', 'versioncontrolsystem', 'vcinteraction',
       'vchostingpersonal use', 'vchostingprofessional use',
       'officestackasynchaveworkedwith', 'officestackasyncwanttoworkwith',
       'officestacksynchavework

In [10]:
# https://cmdlinetips.com/2020/07/cleaning_up_pandas-column-names/#google_vignette

- let's try to understand the more about the data..

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73268 entries, 0 to 73267
Data columns (total 79 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   responseid                      73268 non-null  int64  
 1   mainbranch                      73268 non-null  object 
 2   employment                      71709 non-null  object 
 3   remotework                      58958 non-null  object 
 4   codingactivities                58899 non-null  object 
 5   edlevel                         71571 non-null  object 
 6   learncode                       71580 non-null  object 
 7   learncodeonline                 50685 non-null  object 
 8   learncodecoursescert            29389 non-null  object 
 9   yearscode                       71331 non-null  object 
 10  yearscodepro                    51833 non-null  object 
 11  devtype                         61302 non-null  object 
 12  orgsize                         

#### Observation
  1. Out of 79 columns 73 are object data type
  1. one is integer data type
  1. five is float data type

In [12]:
#let's see how many duplicate values present in our dataset
df.duplicated().sum()

0

**Our data has no duplicated values**

In [33]:
df.dtypes

responseid               int64
mainbranch              object
employment              object
remotework              object
codingactivities        object
                        ...   
truefalse2              object
truefalse3              object
surveylength            object
surveyease              object
convertedcompyearly    float64
Length: 79, dtype: object

In [13]:
df.describe()

responseid     comptotal  vchostingpersonal use  \
count  73268.000000  3.842200e+04                    0.0   
mean   36634.500000  2.342434e+52                    NaN   
std    21150.794099  4.591478e+54                    NaN   
min        1.000000  0.000000e+00                    NaN   
25%    18317.750000  3.000000e+04                    NaN   
50%    36634.500000  7.750000e+04                    NaN   
75%    54951.250000  1.540000e+05                    NaN   
max    73268.000000  9.000000e+56                    NaN   

       vchostingprofessional use       workexp  convertedcompyearly  
count                        0.0  36769.000000         3.807100e+04  
mean                         NaN     10.242378         1.707613e+05  
std                          NaN      8.706850         7.814132e+05  
min                          NaN      0.000000         1.000000e+00  
25%                          NaN      4.000000         3.583200e+04  
50%                          NaN      8.000000         6.784500e+04  
75%                          NaN     15.000000         1.200000e+05  
max                          NaN     50.000000         5.000000e+07

In [14]:
df.describe().T

count          mean           std  min       25%  \
responseid                 73268.0  3.663450e+04  2.115079e+04  1.0  18317.75   
comptotal                  38422.0  2.342434e+52  4.591478e+54  0.0  30000.00   
vchostingpersonal use          0.0           NaN           NaN  NaN       NaN   
vchostingprofessional use      0.0           NaN           NaN  NaN       NaN   
workexp                    36769.0  1.024238e+01  8.706850e+00  0.0      4.00   
convertedcompyearly        38071.0  1.707613e+05  7.814132e+05  1.0  35832.00   

                               50%        75%           max  
responseid                 36634.5   54951.25  7.326800e+04  
comptotal                  77500.0  154000.00  9.000000e+56  
vchostingpersonal use          NaN        NaN           NaN  
vchostingprofessional use      NaN        NaN           NaN  
workexp                        8.0      15.00  5.000000e+01  
convertedcompyearly        67845.0  120000.00  5.000000e+07

In [19]:
df.isnull().sum()

responseid                 0
mainbranch                 0
employment              1559
remotework             14310
codingactivities       14369
                       ...  
truefalse2             37553
truefalse3             37519
surveylength            2824
surveyease              2760
convertedcompyearly    35197
Length: 79, dtype: int64

In [20]:
df[df.columns[df.isnull().any()]]

employment  \
0                                                    NaN   
1                                    Employed, full-time   
2                                    Employed, full-time   
3                                    Employed, full-time   
4                                    Employed, full-time   
...                                                  ...   
73263                                Employed, full-time   
73264                                Employed, full-time   
73265                                Employed, full-time   
73266                                Employed, full-time   
73267  Independent contractor, freelancer, or self-em...   

                                 remotework  \
0                                       NaN   
1                              Fully remote   
2      Hybrid (some remote, some in-person)   
3                              Fully remote   
4      Hybrid (some remote, some in-person)   
...                                     ...   
73263                          Fully remote   
73264                        Full in-person   
73265  Hybrid (some remote, some in-person)   
73266  Hybrid (some remote, some in-person)   
73267                          Fully remote   

                                        codingactivities  \
0                                                    NaN   
1               Hobby;Contribute to open-source projects   
2                                                  Hobby   
3                           I don’t code outside of work   
4                                                  Hobby   
...                                                  ...   
73263                            Freelance/contract work   
73264                                              Hobby   
73265                      Hobby;School or academic work   
73266                                              Hobby   
73267  Hobby;Contribute to open-source projects;Boots...   

                                               edlevel  \
0                                                  NaN   
1                                                  NaN   
2      Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
3         Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4         Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
...                                                ...   
73263     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
73264  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
73265     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
73266     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
73267     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                               learncode  \
0                                                    NaN   
1                                                    NaN   
2      Books / Physical media;Friend or family member...   
3      Books / Physical media;School (i.e., Universit...   
4      Other online resources (e.g., videos, blogs, f...   
...                                                  ...   
73263  Books / Physical media;Other online resources ...   
73264  Other online resources (e.g., videos, blogs, f...   
73265  Books / Physical media;Other online resources ...   
73266         Books / Physical media;On the job training   
73267  Books / Physical media;Friend or family member...   

                                         learncodeonline  \
0                                                    NaN   
1                                                    NaN   
2      Technical documentation;Blogs;Programming Game...   
3                                                    NaN   
4      Technical documentation;Blogs;Stack Overflow;O...   
...                                                  ...   
73263  Technical documentation;Blogs;Written Tutorial...   
73264  Technical documentation;Blogs;Written Tutorial...   
73265  Technical documentation;Programming Games;Stac...   
73266                

In [22]:
df.isna().any()

responseid             False
mainbranch             False
employment              True
remotework              True
codingactivities        True
                       ...  
truefalse2              True
truefalse3              True
surveylength            True
surveyease              True
convertedcompyearly     True
Length: 79, dtype: bool

In [23]:
df[df.columns[df.isna().any()]]

employment  \
0                                                    NaN   
1                                    Employed, full-time   
2                                    Employed, full-time   
3                                    Employed, full-time   
4                                    Employed, full-time   
...                                                  ...   
73263                                Employed, full-time   
73264                                Employed, full-time   
73265                                Employed, full-time   
73266                                Employed, full-time   
73267  Independent contractor, freelancer, or self-em...   

                                 remotework  \
0                                       NaN   
1                              Fully remote   
2      Hybrid (some remote, some in-person)   
3                              Fully remote   
4      Hybrid (some remote, some in-person)   
...                                     ...   
73263                          Fully remote   
73264                        Full in-person   
73265  Hybrid (some remote, some in-person)   
73266  Hybrid (some remote, some in-person)   
73267                          Fully remote   

                                        codingactivities  \
0                                                    NaN   
1               Hobby;Contribute to open-source projects   
2                                                  Hobby   
3                           I don’t code outside of work   
4                                                  Hobby   
...                                                  ...   
73263                            Freelance/contract work   
73264                                              Hobby   
73265                      Hobby;School or academic work   
73266                                              Hobby   
73267  Hobby;Contribute to open-source projects;Boots...   

                                               edlevel  \
0                                                  NaN   
1                                                  NaN   
2      Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
3         Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4         Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
...                                                ...   
73263     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
73264  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
73265     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
73266     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
73267     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                               learncode  \
0                                                    NaN   
1                                                    NaN   
2      Books / Physical media;Friend or family member...   
3      Books / Physical media;School (i.e., Universit...   
4      Other online resources (e.g., videos, blogs, f...   
...                                                  ...   
73263  Books / Physical media;Other online resources ...   
73264  Other online resources (e.g., videos, blogs, f...   
73265  Books / Physical media;Other online resources ...   
73266         Books / Physical media;On the job training   
73267  Books / Physical media;Friend or family member...   

                                         learncodeonline  \
0                                                    NaN   
1                                                    NaN   
2      Technical documentation;Blogs;Programming Game...   
3                                                    NaN   
4      Technical documentation;Blogs;Stack Overflow;O...   
...                                                  ...   
73263  Technical documentation;Blogs;Written Tutorial...   
73264  Technical documentation;Blogs;Written Tutorial...   
73265  Technical documentation;Programming Games;Stac...   
73266                

In [25]:
# saprate catigorical and numerical features
catigorical_features = df.select_dtypes(include="object").columns
numerical_features = df.select_dtypes(exclude="object").columns

In [26]:
catigorical_features

Index(['mainbranch', 'employment', 'remotework', 'codingactivities', 'edlevel',
       'learncode', 'learncodeonline', 'learncodecoursescert', 'yearscode',
       'yearscodepro', 'devtype', 'orgsize', 'purchaseinfluence', 'buynewtool',
       'country', 'currency', 'compfreq', 'languagehaveworkedwith',
       'languagewanttoworkwith', 'databasehaveworkedwith',
       'databasewanttoworkwith', 'platformhaveworkedwith',
       'platformwanttoworkwith', 'webframehaveworkedwith',
       'webframewanttoworkwith', 'misctechhaveworkedwith',
       'misctechwanttoworkwith', 'toolstechhaveworkedwith',
       'toolstechwanttoworkwith', 'newcollabtoolshaveworkedwith',
       'newcollabtoolswanttoworkwith', 'opsysprofessional use',
       'opsyspersonal use', 'versioncontrolsystem', 'vcinteraction',
       'officestackasynchaveworkedwith', 'officestackasyncwanttoworkwith',
       'officestacksynchaveworkedwith', 'officestacksyncwanttoworkwith',
       'blockchain', 'newsosites', 'sovisitfreq', 'so

In [28]:
catigorical_features.shape

(73,)

In [29]:
numerical_features

Index(['responseid', 'comptotal', 'vchostingpersonal use',
       'vchostingprofessional use', 'workexp', 'convertedcompyearly'],
      dtype='object')

In [30]:
numerical_features.shape

(6,)

In [31]:
# fetch the unique value from catigorical data
for i in catigorical_features:
    print(i,df[i].unique())
    
    print()
    print("*****************************Seperated_Columns********************************")
    print()

mainbranch ['None of these' 'I am a developer by profession'
 'I am not primarily a developer, but I write code sometimes as part of my work'
 'I code primarily as a hobby' 'I am learning to code'
 'I used to be a developer by profession, but no longer am']

*****************************Seperated_Columns********************************

employment [nan 'Employed, full-time' 'Student, full-time' 'Student, part-time'
 'Not employed, but looking for work'
 'Independent contractor, freelancer, or self-employed'
 'Employed, full-time;Independent contractor, freelancer, or self-employed'
 'Employed, part-time' 'Student, part-time;Employed, part-time'
 'Not employed, and not looking for work'
 'Student, full-time;Employed, part-time'
 'Employed, full-time;Student, part-time'
 'Employed, full-time;Student, full-time'
 'Student, part-time;Independent contractor, freelancer, or self-employed'
 'Retired' 'Student, full-time;Not employed, but looking for work'
 'I prefer not to say'
 'Student, ful

toolstechhaveworkedwith [nan 'npm' 'Homebrew' 'Homebrew;npm' 'Docker;npm;Terraform'
 'Docker;Homebrew;Kubernetes' 'Docker' 'Docker;Homebrew;npm;Yarn'
 'Docker;npm;Yarn' 'Docker;npm' 'Docker;Kubernetes'
 'Docker;Kubernetes;npm;Terraform;Yarn' 'Kubernetes;npm;Terraform'
 'Docker;Kubernetes;npm' 'Ansible;Docker;npm;Terraform'
 'Ansible;Docker;Terraform' 'npm;Yarn'
 'Docker;Homebrew;Kubernetes;npm;Yarn' 'Ansible;Docker;npm'
 'Docker;Homebrew;Unity 3D'
 'Ansible;Docker;Homebrew;Kubernetes;npm;Puppet;Terraform'
 'Ansible;Docker;Homebrew;Kubernetes;Terraform'
 'Docker;Flow;Kubernetes;npm;Unity 3D' 'Docker;Homebrew'
 'Ansible;Docker;Kubernetes;npm;Terraform;Unity 3D;Unreal Engine;Yarn'
 'Docker;Puppet' 'Docker;Homebrew;npm' 'Unity 3D'
 'Docker;Kubernetes;Puppet' 'Docker;Kubernetes;npm;Yarn'
 'Docker;npm;Terraform;Yarn' 'Docker;Yarn' 'Ansible;Docker;Kubernetes'
 'Docker;Terraform' 'Ansible;Docker;Kubernetes;npm;Terraform;Yarn'
 'Kubernetes;npm' 'Docker;Homebrew;Yarn' 'Ansible;Chef;Docker;Terraf

In [32]:
for i in numerical_features:
    print(i,df[i].unique())
    
    print()
    print("*****************************Seperated_Columns********************************")
    print()

responseid [    1     2     3 ... 73266 73267 73268]

*****************************Seperated_Columns********************************

comptotal [     nan   32000.   60000. ... 1038000.   64200.  439000.]

*****************************Seperated_Columns********************************

vchostingpersonal use [nan]

*****************************Seperated_Columns********************************

vchostingprofessional use [nan]

*****************************Seperated_Columns********************************

workexp [nan  6. 14.  5. 15.  4. 23.  9. 22. 21.  3. 28.  7. 12.  8.  2. 20. 27.
  1. 10. 19. 13. 30. 11. 18. 47. 17. 46. 25. 24. 31.  0. 41. 43. 40. 16.
 42. 35. 32. 34. 26. 36. 38. 29. 33. 44. 37. 50. 45. 39. 48. 49.]

*****************************Seperated_Columns********************************

convertedcompyearly [    nan  40205. 215232. ...  16203.  43324. 110245.]

*****************************Seperated_Columns********************************



In [35]:
#Let's create some new features with help of existing feature:
features =[]
dtypes =[]
count =[]
uniques =[]
missing = []
missing_percentage=[]

for j in df.columns:
    features.append(j)
    count.append(len(df[j]))
    missing.append(df[j].isnull().sum())
    uniques.append(df[j].nunique())
    dtypes.append(df[j].dtypes)
    missing_percentage.append((df[j].isnull().sum()/df.shape[0])*100)
    
dataframe = pd.DataFrame({'feature':features,
                         'count':count,
                         'missing':missing,
                         'nunique':uniques,
                         'dtypes':dtypes,
                         'missing_percentage':missing_percentage})

dataframe.set_index('feature')

count  missing  nunique   dtypes  missing_percentage
feature                                                                  
responseid           73268        0    73268    int64            0.000000
mainbranch           73268        0        6   object            0.000000
employment           73268     1559      103   object            2.127805
remotework           73268    14310        3   object           19.531037
codingactivities     73268    14369       63   object           19.611563
...                    ...      ...      ...      ...                 ...
truefalse2           73268    37553        2   object           51.254299
truefalse3           73268    37519        2   object           51.207894
surveylength         73268     2824        3   object            3.854343
surveyease           73268     2760        3   object            3.766992
convertedcompyearly  73268    35197     7909  float64           48.038707

[79 rows x 5 columns]

In [37]:
for j in catigorical_features:
    print(j,df[j].value_counts())
    print()
    print('-----------------------------------------------------------------------------------')
    print()

mainbranch I am a developer by profession                                                   53507
I am learning to code                                                             6309
I am not primarily a developer, but I write code sometimes as part of my work     5794
I code primarily as a hobby                                                       4865
None of these                                                                     1497
I used to be a developer by profession, but no longer am                          1296
Name: mainbranch, dtype: int64

-----------------------------------------------------------------------------------

employment Employed, full-time                                                                                                                    42962
Student, full-time                                                                                                                      6756
Independent contractor, freelancer, or self-employed     

misctechwanttoworkwith .NET                                                                                                                                                                                                             5025
Flutter                                                                                                                                                                                                          1740
Spring                                                                                                                                                                                                           1658
React Native                                                                                                                                                                                                     1231
Apache Kafka                                                                                                             

frequency1 1-2 times a week     21689
Never                 8754
3-5 times a week      3489
6-10 times a week      796
10+ times a week       643
Name: frequency1, dtype: int64

-----------------------------------------------------------------------------------

frequency2 1-2 times a week     16368
3-5 times a week      8146
10+ times a week      3950
6-10 times a week     3454
Never                 3426
Name: frequency2, dtype: int64

-----------------------------------------------------------------------------------

frequency3 1-2 times a week     17666
Never                11081
3-5 times a week      4019
6-10 times a week      969
10+ times a week       780
Name: frequency3, dtype: int64

-----------------------------------------------------------------------------------

timesearching 30-60 minutes a day           13652
15-30 minutes a day           10122
60-120 minutes a day           6371
Less than 15 minutes a day     3528
Over 120 minutes a day         2525
Name: timesearchi